### ISSUE TITLE
Narrative Interface

### ISSUE THREAD

**Issue by jthrilly** [codaco/Network-Canvas#258]
Convex hulls: https://ptsjs.org/demo/?name=polygon.convexHull

**Comment by rebeccamadsen** [comment]
I see this generally as an extension of Sociogram, and have seen mockups for this, but lack some context on what the specific dropdowns are for: Presets, Highlighted, Links, Contexts.

![05network_canvas_sociogram_narrative-open_16x9](https://user-images.githubusercontent.com/16093053/48784820-6e60bc00-ecb1-11e8-8a89-07d09672a4c3.jpg)

I also see a "Change" and "Reset" button and a toggle button. Is "reset" for reseting the view? Does a "Change" button imply that any toggles in the dropdowns are not immediate? What is the toggle button for?

How are convex hulls related?

**Comment by jthrilly** [comment]
This is an extension of sociogram - exactly. Possibly the best way of understanding it is to look at the prototype, which I think you have access to? In case you don't, I'll put up a few GIFs/PNGs below.

The main features of the interface are twofold:

- To allow quick switching between different "presets" of node and edge display properties.
- To allow annotation features such as drawing and highlighting.

> I also see a "Change" and "Reset" button and a toggle button. Is "reset" for reseting the view? Does a "Change" button imply that any toggles in the dropdowns are not immediate? What is the toggle button for?

- This panel was really just a very vague mockup. The intention is that the panel functions like a key on a chart. It can be selectively shown in order to show the user which variables are mapped on to the graph in the current preset.
- The "change" button was intended to be the way the user changes the preset. I don't think we need it.
- The reset button immediately removes any on screen annotations, and resets any interactive node state.
- The toggle was for showing/hiding convex hulls.

> How are convex hulls related?

Convex hulls are a way of showing (overlapping) group membership of nodes. Any categorical variable can potentially be used as a data source for displaying convex hulls. Here's a screenshot:

<img width="1680" alt="screenshot 2018-11-20 at 11 02 17" src="https://user-images.githubusercontent.com/1387940/48836229-2dd96f00-ed93-11e8-96fc-68c7cdd11c92.png">


# Functionality

So functionality wise, here is what I think we are looking at:
- "Presets" - collections of node/edge display properties, and behaviour settings.
- "Behaviours"
  - Free drawing on the screen. The drawings should fade out over time (I did this proportional to their length, which worked pretty well)
  - "Highlighting" - This is different from nodes being in a selected state, and is a temporary way of visually indicating which node is being discussed. We will need to think about how we do this visually.
  - "Convex hulls" - Visually displaying group membership, based on a categorical variable.

Here's kind of what I am thinking for the API:

```
{
  "id": "narrative",
  "type": "Narrative",
  "label": "Narrative",
  "presets": {
    "preset1": {
      "label": "My preset",
      "subject": {
        "entity": "node",
        "type": "4aebf73e-95e3-4fd1-95e7-237dcc4a4466"
      },
      "layoutVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
      "groupVariable": "d13ca72d-aefe-4f48-841d-09f020e0e988",
      "edges": {
        "display": ["77199445-9d50-4646-b0bc-6d6b0c0e06bd", "867127d9-086b-403a-a3e7-2c7d32126546"]
      },
      "highlight": [
        {
          "variable": "03b03617-46ae-41cb-9462-9acd8a17edd6",
          "color": "node-highlight-seq-1"
        },
        {
          "variable": "03b03617-46ae-41cb-9462-9acd8a17edd6",
          "color": "node-highlight-seq-2"
        }
      ]
    }
  },
  "behaviours": {
    "freeDraw": true,
    "lasso": true,
    "featureNode": true,
    "allowRepositioning": true
  }
}
```

**Comment by jthrilly** [comment]
Given the complexity, it might be best if we have a quick call to discuss how we will divide this one up?

**Comment by wwqrd** [comment]
After some discussion about this, we're thinking we could separate this interface out into sub-tasks, I'm just writing up the issues for those now, will post back here when they're done.

**Comment by wwqrd** [comment]
This interface can be split into the following distinct tasks:

- [Narrative Interface: Convex hulls layer component](#797)
- [Narrative Interface: Annotations layer component](#799)
- [Narrative Interface: Control panel component](#800)

Control panel is slightly dependent on #797 and #799, but to some extent it can still be completed without those.

We'll need another task to weave them all together.

**Comment by wwqrd** [comment]
I've also added an issue to create the Narrative interface, so that anyone taking on the sub-components has somewhere consistent to put things. https://github.com/codaco/Network-Canvas/issues/801 - So this one probably needs to happen first.

**Comment by rebeccamadsen** [comment]
Some API questions:
* This API does not have a "background" like sociogram. Should it?
* What is a "featureNode" behavior?
* Is "groupVariable" the categorical variable indicating which convex hulls to display?
* Are "lasso" and "freeDraw" separate, or simultaneous options? How complex of a lasso are we intending, and what would the lasso-ed area do?
* layout in sociogram has an "allowPositioning" option, but here you've placed "allowRepositioning" as a behavior. Is that separate from the layout for a reason? Is it interface specific and not preset specific?
* I assume we aren't toggling highlight states for node attributes, and not allowing edges to be created, based on this API. Is that accurate?
* Should presets be an array? Does order matter? Are presets tied to the timeline or is it completely exploratory?

**Comment by wwqrd** [comment]
> This API does not have a "background" like sociogram. Should it?

Probably!

> What is a "featureNode" behavior?

I could only speculate 😄 @jthrilly?

> Is "groupVariable" the categorical variable indicating which convex hulls to display?

That's my understanding

> Are "lasso" and "freeDraw" separate, or simultaneous options? How complex of a lasso are we intending, and what would the lasso-ed area do?

Freedraw will be the annotations interface. Lasso, I don't know @jthrilly? 

> layout in sociogram has an "allowPositioning" option, but here you've placed "allowRepositioning" as a behavior. Is that separate from the layout for a reason? Is it interface specific and not preset specific?

Positioning isn't persisted to the network state for this interface. Do we need to have a new property name for that? ...It does keep them conceptually separate.

> I assume we aren't toggling highlight states for node attributes, and not allowing edges to be created, based on this API. Is that accurate?

I think that's correct. We will be toggling which attributes are highlighted, but the participant won't be able to change the attributes. At least for the first pass.

> Should presets be an array? Does order matter? Are presets tied to the timeline or is it completely exploratory?

I think they should be an array too. I don't believe they are tied to to timeline, but there will be a default preset for the initial state.

**Comment by jthrilly** [comment]
Yeesh. Sorry everyone. 

> * This API does not have a "background" like sociogram. Should it?

Yes! Whoops.

> * What is a "featureNode" behavior?

Feature node is designed to be a visual indication that a node or group of nodes are being talked about. It is a tap interaction, which replaces the highlight interaction on the sociogram. So whereas on sociogram tapping a node would toggle a property (and the highlight parameter), on narrative interface, tapping a node will 'feature' it. 

The feature state of a node is reset by the reset button.

In terms of what it should look like...I was thinking some sort of pulsing scale animation that loops indefinitely. 

> * Is "groupVariable" the categorical variable indicating which convex hulls to display?

Yes. 

> * Are "lasso" and "freeDraw" separate, or simultaneous options? How complex of a lasso are we intending, and what would the lasso-ed area do?

Lasso is a pipe dream of mine. The idea would be that you could encircle a group of nodes, and they would then enter the featured state (see above). I think this is probably a bit complicated for a first pass, so let's just focus on freeDraw.

> * layout in sociogram has an "allowPositioning" option, but here you've placed "allowRepositioning" as a behavior. Is that separate from the layout for a reason? Is it interface specific and not preset specific?

My thinking is that everything in behaviours is interface wide. Presets are a less "strong" concept than prompts, and so can share behaviours.

> * I assume we aren't toggling highlight states for node attributes, and not allowing edges to be created, based on this API. Is that accurate?

Yes. This interface _does not_ create or change any node or edge data (apart from layout).

> * Should presets be an array? Does order matter? Are presets tied to the timeline or is it completely exploratory?

Great questions!

Presets aren't tied to the timeline. This is a single interface, so one point on the timeline. This is purely exploratory.

I think order does matter, since the researcher will probably want to start somewhere specific.

**Comment by jthrilly** [comment]
That's a great spot Rebecca. Let's move the subject definition to the root level.

**Comment by rebeccamadsen** [comment]
Resolved in #802.
